crates/bevy_ecs/src/change_detection.rs

```diff
@@ -0,0 +1,152 @@
+use crate::component::{Component, ComponentTicks};
+use bevy_reflect::Reflect;
+use std::ops::{Deref, DerefMut};
+
+pub trait ChangeDetectable {
+    /// Returns true if (and only if) this value been added since the last execution of this
+    /// system.
+    fn is_added(&self) -> bool;
+
+    /// Returns true if (and only if) this value been changed since the last execution of this
+    /// system.
+    fn is_changed(&self) -> bool;
+
+    /// Manually flags this value as having been changed. This normally isn't
+    /// required because accessing this pointer mutably automatically flags this
+    /// value as "changed".
+    ///
+    /// **Note**: This operation is irreversible.
+    fn set_changed(&mut self);
+}
+
+macro_rules! change_detection_impl {
+    ($name:ident < $( $generics:tt ),+ >, $target:ty, $($traits:ident)?) => {
+        impl<$($generics),* $(: $traits)?> ChangeDetectable for $name<$($generics),*>
+        {
+            #[inline]
+            fn is_added(&self) -> bool {
+                self.ticks
+                    .component_ticks
+                    .is_added(self.ticks.last_change_tick, self.ticks.change_tick)
+            }
+
+            #[inline]
+            fn is_changed(&self) -> bool {
+                self.ticks
+                    .component_ticks
+                    .is_changed(self.ticks.last_change_tick, self.ticks.change_tick)
+            }
+
+            #[inline]
+            fn set_changed(&mut self) {
+                self.ticks
+                    .component_ticks
+                    .set_changed(self.ticks.change_tick);
+            }
+        }
+
+        impl<$($generics),* $(: $traits)?> Deref for $name<$($generics),*>
+        {
+            type Target = $target;
+
+            #[inline]
+            fn deref(&self) -> &Self::Target {
+                self.value
+            }
+        }
+
+        impl<$($generics),* $(: $traits)?> DerefMut for $name<$($generics),*>
+        {
+            #[inline]
+            fn deref_mut(&mut self) -> &mut Self::Target {
+                self.set_changed();
+                self.value
+            }
+        }
+
+        impl<$($generics),* $(: $traits)?> AsRef<$target> for $name<$($generics),*>
+        {
+            #[inline]
+            fn as_ref(&self) -> &$target {
+                self.deref()
+            }
+        }
+
+        impl<$($generics),* $(: $traits)?> AsMut<$target> for $name<$($generics),*>
+        {
+            #[inline]
+            fn as_mut(&mut self) -> &mut $target {
+                self.deref_mut()
+            }
+        }
+    };
+}
+
+macro_rules! impl_into_inner {
+    ($name:ident < $( $generics:tt ),+ >, $target:ty, $($traits:ident)?) => {
+        impl<$($generics),* $(: $traits)?> $name<$($generics),*>
+        {
+            /// Consume `self` and return the contained value while marking `self` as "changed".
```

Comment:
```suggestion
            /// Consume `self` and return a mutable reference to the contained value while marking `self` as "changed".
```

This is an important distinction, as the mutable reference from the container is almost always why you want to use this method.